src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/CompilerTypeSystemContext.ValueTypeMethods.cs

```diff
@@ -73,6 +75,16 @@ private bool RequiresGetFieldHelperMethod(MetadataType valueType)
             if (valueType.IsWellKnownType(WellKnownType.Double) || valueType.IsWellKnownType(WellKnownType.Single))
                 return false;
 
+            // Heuristic: async state machines don't need equality/hashcode.
+            _iAsyncStateMachineType ??= SystemModule.GetType("System.Runtime.CompilerServices", "IAsyncStateMachine");
+            if (valueType.HasCustomAttribute("System.Runtime.CompilerServices", "CompilerGeneratedAttribute")
+                && valueType.ContainingType != null
```

Comment:
I do not think we need to this check for nested type. `CompilerGeneratedAttribute` + implements `IAsyncStateMachine` should be enough.

Reply:
I'm worried about this kicking in for wrong types. But agreed that this doesn't prevent that much...

Reply:
Can we make the failure mode in case this kicks-in unexpectedly explicit? E.g. set Equals and GetHashCode slots to null, so that the call of these methods hits null reference exception; or set the slots to point to a helper that throws unconditionally.